Tradução de Async & Performace - Cap. 2 - seção: Doing Versus Planning (#339)
cez-aug/You-Dont-Know-JS: async & performance/ch2.md

Eu não sou interrompido e puxado para outro "processo" em todas as oportunidades que eu poderia ter (felizmente - ou esse livro nunca seria escrito!). Mas isso acontece frequente o suficiente para que eu sinta que meu próprio cérebro está quase constantemente mudando para vários contextos diferentes (também conhecidos como "processos"). E é muito parecido com o que o mecanismo JS provavelmente sentiria.

### Fazendo Versus Planejando

OK, então nossos cérebros podem ser pensados como operando na fila do loop do eventos de thread única, assim como o mecanismo JS. Isso soa como uma boa combinação.

Mas precisamos ter mais nuances do que isso em nossa análise. Há uma grande e observável diferença entre como planejamos várias tarefas e como nosso cérebro realmente as opera.

> Novamente, voltando à redação deste texto como minha metáfora. Meu plano básico de esboço mental aqui é continuar escrevendo e escrevendo, passando sequencialmente por um conjunto de pontos que ordenei em meus pensamentos. Não pretendo ter nenhuma interrupção ou atividade não linear neste trabalho. Mas, no entanto, meu cérebro está mudando o tempo todo.

Reviewer: O que acha dessas alterações?
```suggestion
Novamente, voltando à redação deste texto como minha metáfora. O esboço grosseiro do meu plano aqui é continuar escrevendo e escrevendo, passando sequencialmente por um conjunto de pontos que ordenei em meus pensamentos. Não pretendo ter nenhuma interrupção ou atividade não linear neste trabalho. Mas mesmo assim, meu cérebro está fazendo rodízio o tempo todo.
```

Embora em um nível operacional nossos cérebros sejam assíncronos, parecemos planejar tarefas de maneira sequencial e síncrona. "Eu preciso ir à loja, comprar um pouco de leite e depois deixar minha lavagem a seco."

> Você notará que esse pensamento de nível superior (planejamento) não parece muito assíncrono em sua formulação. De fato, é meio raro pensarmos deliberadamente apenas em termos de eventos. Em vez disso, planejamos as coisas com cuidado, sequencialmente (A então B e depois C), e assumimos, de certa forma, uma espécie de bloqueio temporal que força B a esperar A e C a esperar B.

Reviewer: ```suggestion
Você notará que esse pensamento (planejamento) de alto nível não parece muito com eventos assíncronos em sua formulação. De fato, é meio raro pensarmos deliberadamente apenas em termos de eventos. Em vez disso, planejamos as coisas com cuidado, sequencialmente (A então B e depois C), e assumimos, de certa forma, uma espécie de bloqueio temporal que força B a esperar A e C a esperar B.
```

> Quando um desenvolvedor escreve código, ele está planejando um conjunto de ações para ocorrer. Se eles são bons em ser desenvolvedor, estão **planejando cuidadosamente.** "Eu preciso definir `z` para o valor de `x` e depois `x` para o valor de `y`," e assim por diante.

Reviewer: A frase em inglês tá meio sem concordância de número. O que acha de deixarmos padronizado pra usar plural?
```suggestion
Quando desenvolvedores escrevem código, eles estão planejando a execução de um conjunto de ações. Se eles são bons em ser desenvolvedores, estão **planejando cuidadosamente**. "Eu preciso definir `z` para o valor de `x` e depois `x` para o valor de `y`," e assim por diante.
```

Quando escrevemos código síncrono, declaração por declaração, ele funciona muito como nossa lista de tarefas a fazer:

```js
// troca `x` por `y` (através da varável temporária `z`)
z = x;
x = y;
y = z;
```

> Essas três instruções de atribuição são síncronas, portanto, `x = y` espera que `z = x` termine, e `y = z`, por sua vez, espera que `x = y` termine. Outra maneira de dizer é que essas três instruções são temporariamente obrigadas a executar em uma determinada ordem, uma logo após a outra. Felizmente, não precisamos nos preocupar com detalhes de eventos assíncronos aqui. Se o fizermos, o código fica muito mais complexo, rapidamente!

Reviewer: ```suggestion
Essas três instruções de atribuição são síncronas, portanto `x = y` espera que `z = x` termine e `y = z`, por sua vez, espera que `x = y` termine. Outra maneira de dizer é que essas três instruções são temporariamente obrigadas a executar em uma determinada ordem, uma logo após a outra. Felizmente, não precisamos nos preocupar com detalhes de eventos assíncronos aqui. Se o fizermos, o código fica muito mais complexo, rapidamente!
```

Portanto, se o planejamento do cérebro síncrono mapeia bem as instruções de código síncrono, quão bem nossos cérebros se saem planejando o código assíncrono?

> Acontece que a maneira como expressamos assincronia (com retornos de chamada) em nosso código não é muito boa para esse comportamento de planejamento do cérebro síncrono.

Reviewer: ```suggestion
Acontece que a maneira como expressamos assincronia (com callbacks) em nosso código não é muito boa para essa maneira de planejar do cérebro síncrono.
```

Você pode realmente imaginar ter uma linha de pensamento que planeje suas tarefas assim?

> > "Eu preciso ir à loja, mas no caminho tenho certeza de que vou receber um telefonema, então 'Oi, mãe', e enquanto ela começa a falar, procurarei o endereço da loja no GPS, mas isso levará um segundo para carregar, então eu vou desligar o rádio para ouvir melhor mamãe, depois vou perceber que esqueci de vestir uma jaqueta e está frio lá fora, mas não importa, continuo dirigindo e conversando com mamãe e, em seguida, o toque do cinto de segurança me lembra de apertar o cinto, então 'Sim, mãe, estou usando meu cinto de segurança, sempre uso!'. Ah, finalmente o GPS recebeu as instruções, agora..."

Reviewer: `diff` no trecho do cinto de segurança:
```suggestion
> "Eu preciso ir à loja, mas no caminho tenho certeza de que vou receber um telefonema, então 'Oi, mãe', e enquanto ela começa a falar, procurarei o endereço da loja no GPS, mas isso levará um segundo para carregar, então eu vou desligar o rádio para ouvir melhor mamãe, depois vou perceber que esqueci de vestir uma jaqueta e está frio lá fora, mas não importa, continuo dirigindo e conversando com mamãe e, em seguida, o toque do cinto de segurança me lembra de colocá-lo, então 'Sim, mãe, estou usando meu cinto de segurança, sempre uso!'. Ah, finalmente o GPS recebeu as instruções, agora..."
```

Por mais ridículo que pareça uma formulação de como planejamos nosso dia e pensamos sobre o que fazer e em que ordem, é exatamente dessa forma como nosso cérebro opera em um nível funcional. Lembre-se, isso não é multitarefa, é apenas uma mudança rápida de contexto.

> A razão pela qual é difícil para nós, como desenvolvedores, escrever código assíncrono, especialmente quando tudo o que temos é o retorno de chamadas (callback), é que o fluxo de pensamento/planejamento consciente não é natural para a maioria de nós.

Reviewer: ```suggestion
A razão pela qual é difícil para nós, como desenvolvedores, escrever código assíncrono, especialmente quando tudo o que temos são callback, é que o fluxo de pensamento/planejamento consciente não é natural para a maioria de nós.
```

> Nós pensamos em termos sequenciais, mas as ferramentas (retornos de chamada) disponíveis no código não são expressas de uma maneira gradual (passo a passo) uma vez que passamos de síncrono para assíncrono.

Reviewer: ```suggestion
Nós pensamos em termos sequenciais, mas as ferramentas disponíveis no código (callbacks) não são expressas de uma maneira sequencial uma vez que passamos de síncrono para assíncrono.
```

> E é por **isso** que é tão difícil escrever e raciocinar com precisão sobre o código JS assíncrono com retornos de chamada: porque não é assim que o nosso planejamento cerebral funciona.

Reviewer: ```suggestion
E é por **isso** que é tão difícil escrever e raciocinar com precisão sobre o código JS assíncrono com callbacks: porque não é assim que o nosso planejamento cerebral funciona.
```

> **Nota:** A única coisa pior do que não saber por que algum código quebra é não saber por que ele funcionou em primeiro lugar! É a mentalidade clássica do "castelo de cartas": "funciona, mas não sei por quê, então ninguém toca!" Você pode ter ouvido: "O inferno são os outros" (Sartre), e o meme da reviravolta do programador: "O inferno é o código de outras pessoas". Eu acredito verdadeiramente que: "O inferno é não está entendendo meu próprio código". E os retornos de chamada são um dos principais culpados.

Reviewer: ```suggestion
**Nota:** A única coisa pior do que não saber por que algum código quebra é, primeiramente, não saber por que ele funcionou! É a mentalidade clássica do "castelo de cartas": "funciona, mas não sei por quê, então ninguém toca!" Você pode ter ouvido: "O inferno são os outros" (Sartre), e o meme da reviravolta do programador: "O inferno é o código de outras pessoas". Eu acredito verdadeiramente que: "O inferno é não entender meu próprio código". E os callbacks são um dos principais culpados.
```
